### PR TITLE
[issue #287] adds Arquillian Recorder Reporter integration

### PR DESCRIPTION
**Pull request by lordofthejars**
This change is [Reviewable](https://reviewable.io/reviews/arquillian/arquillian-cube/412)

### CODE REVIEW

File: core/src/main/java/org/arquillian/cube/impl/reporter/TakeCubeInformation.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.jboss.arquillian.core.api.Event;
+import org.jboss.arquillian.core.api.annotation.Inject;
+import org.jboss.arquillian.core.api.annotation.Observes;
+
+import java.util.List;
+
+/**
+ * Observer that is executed before suite and prints all metrics reported by registered cubes.
+ */
+public class TakeCubeInformation {
```
Comment (bartoszmajsak) [inline core/src/main/java/org/arquillian/cube/impl/reporter/TakeCubeInformation.java]: I think we can find a better name :)

File: core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java [hunk 2]
```diff
@@ -31,17 +32,22 @@ public void register(ExtensionBuilder builder) {
...
         // Only register if container-test-impl is on classpath
-        if(Validate.classExists("org.jboss.arquillian.container.test.impl.enricher.resource.OperatesOnDeploymentAwareProvider")) {
+        if (Validate.classExists("org.jboss.arquillian.container.test.impl.enricher.resource.OperatesOnDeploymentAwareProvider")) {
             builder.service(ResourceProvider.class, CubeIDResourceProvider.class);
         }
+
+        // Only if recorder-reporter is in classpath we should provide reporting capabilities.
+        if (Validate.classExists("org.arquillian.recorder.reporter.ReporterExtension")) {
+            builder.observer(TakeCubeInformation.class);
+        }
```
Comment (bartoszmajsak) [inline core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java]: As a general thought - I'm not sure if this way of validation is the best one. If we change package names or shuffle classes around, or even change their names this will be all broken. We should think about some other way of detection in the nearest future.

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        this.dockerCube = dockerCube;
+    }
+
+    @Override
+    public Reportable report() {
+
+        GroupEntry groupEntry = new GroupEntry(dockerCube.getId());
+
+        boolean error;
+        switch(dockerCube.state()) {
```
Comment (bartoszmajsak) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java]: If that's an enum we can use `EnumSet.of(....).contains(dockerCube.state())` instead of switch.

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java [hunk 4]
```diff
@@ -139,10 +155,14 @@ public void stop() throws CubeControlException {
...
         try {
             lifecycle.fire(new BeforeStop(id));
+
+            long currentTime = System.currentTimeMillis();
             try {
                 executor.stopContainer(id);
             } catch(NotFoundException e) {
             } catch (NotModifiedException e) {}
+            this.stoppingTimeInMillis = System.currentTimeMillis() - currentTime;
+
```
Comment (bartoszmajsak) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java]: Just a thought - can't we collect metrics through events instead of keeping them as a part of the object?

File: docker/drone/src/main/java/org/arquillian/cube/docker/drone/util/VideofileDestination.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class VideoFileDestination {
+
+    private static final File DEFAULT_LOCATION_OUTPUT_MAVEN = new File("target");
+    private static final File DEFAULT_LOCATION_OUTPUT_GRADLE = new File("build");
+
+    private static final File MAVEN_REPORT_DIR = new File(DEFAULT_LOCATION_OUTPUT_MAVEN, "surefire-reports");
```
Comment (bartoszmajsak) [inline docker/drone/src/main/java/org/arquillian/cube/docker/drone/util/VideofileDestination.java]: What if we run tests using failsafe?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Class that reports generic Docker information like orchestration or docker version.
+ */
+public class TakeDockerEnvironment {
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Why `Take`? Does it really explain what is the responsibility for the test?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            screenshotEntry.setWidth(bufferedImage.getWidth());
+            screenshotEntry.setHeight(bufferedImage.getHeight());
+            screenshotEntry.setSize(String.valueOf(imageFile.length()));
+            screenshotEntry.setMessage("Docker Composition before executing tests.");
+
+            return screenshotEntry;
+        } catch (IOException e) {
+            log.log(Level.WARNING, String.format("Docker compositions schema could not be generated because of %s.", e));
+        }
+
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: I don't know the API here but shouldn't we close `bufferedImage` here?

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        return null;
```
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Maybe some null object instead of just `null`?

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        final PropertyEntry propertyDockerInfoEntry = rootEntries.get(0);
+        assertThat(propertyDockerInfoEntry).isInstanceOf(GroupEntry.class);
+
+        GroupEntry dockerInfoEntry = (GroupEntry) propertyDockerInfoEntry;
+        final List<PropertyEntry> propertyEntries = dockerInfoEntry.getPropertyEntries();
+
+        assertThat(propertyEntries).hasSize(5);
+
+        assertThat(propertyEntries.get(0)).isEqualToComparingFieldByField(new KeyValueEntry("Version", "1.1.0"));
+        assertThat(propertyEntries.get(1)).isEqualToComparingFieldByField(new KeyValueEntry("OS", "linux"));
```
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Don't we have `equals` for `KeyValueEntry`? If so - we can use assertion for lists, something list `containsExactly`. Then we don't even have to assert the size.

File: core/src/main/java/org/arquillian/cube/impl/reporter/TakeCubeInformation.java (hunk 1 above)
Comment (lordofthejars) [inline core/src/main/java/org/arquillian/cube/impl/reporter/TakeCubeInformation.java]: Basically I followed the same convention used in Recorder integrations. They use this Take + <blabla>. We can change it but this is the reason why I used this name

File: core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java (hunk 2 above)
Comment (lordofthejars) [inline core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java]: Yes but it is used all over Arquillian project.https://github.com/arquillian/arquillian-core/blob/c2ae438796b4bab56a86c6246982ecfe038f4a78/core/spi/src/main/java/org/jboss/arquillian/core/spi/LoadableExtension.java#L46 What we can do is add a new method that receives a Class object and inside to the getFullQualifiedName thing an call the String one, so at least we are not breaking anything and in future we can use the Class approach.

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java (hunk 3 above)
Comment (lordofthejars) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java]: But then we will have a list of if/else

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java (hunk 4 above)
Comment (lordofthejars) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java]: I thought the same but notice that there are some before/after events. So if we took there it means that we could add to the time, the duration of the execution of events thrown there. So I decided that the most real time was doing in that way, although as you said it seems event would be better. WDYT?

File: docker/drone/src/main/java/org/arquillian/cube/docker/drone/util/VideofileDestination.java (hunk 5 above)
Comment (lordofthejars) [inline docker/drone/src/main/java/org/arquillian/cube/docker/drone/util/VideofileDestination.java]: Good point that I didn't take into consideration.

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java (hunk 6 above)
Comment (lordofthejars) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Yes just to follow same convention used in other recorder integrations (graphene, drone, ...)

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 9 above)
Comment (lordofthejars) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: No sadly the Recorder entries has no equals implemented. We can do it as a separate PR in Recorder, for this reason now it is in this way.

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java (hunk 7 above)
Comment (lordofthejars) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Yes I think try with resources would be the best way here.

Comment (lordofthejars) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: BufferedImage has no close method nor implements `AutoClosable` so it is not required

File: core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java (hunk 2 above)
Comment (bartoszmajsak) [inline core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java]: Let's think about it. Shall we creat a jira issue/card for it so we won't forget to address it?

File: core/src/main/java/org/arquillian/cube/impl/reporter/TakeCubeInformation.java (hunk 1 above)
Comment (bartoszmajsak) [inline core/src/main/java/org/arquillian/cube/impl/reporter/TakeCubeInformation.java]: Ok, clear.

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java (hunk 3 above)
Comment (bartoszmajsak) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java]: I don't see where you would need an if-else. You simply do something like

```java
error = EnumSet.of(START_FAILED, CREATE_FAILED, STOP_FAILED, DESTORY_FAILED)
               .contains(dockerCube.state());
```

or am I missing something?

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java (hunk 4 above)
Comment (bartoszmajsak) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java]: Fair point. Not sure what is best. I think I have to spend a bit more time reading the code to make better justification. If you feel like this is good way to go I'm fine with that.

File: docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java (hunk 7 above)
Comment (bartoszmajsak) [inline docker/recorder/src/main/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironment.java]: Ok clear, I was just looking at the code here, not anywhere from the IDE.

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 9 above)
Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Then maybe we should bring them? WDYT?

File: core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java (hunk 2 above)
Comment (lordofthejars) [inline core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java]: We should add a card as community tasks

Comment (bartoszmajsak) [inline core/src/main/java/org/arquillian/cube/impl/client/CubeExtension.java]: use 'help wanted' label then

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java (hunk 3 above)
Comment (lordofthejars) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/client/metadata/ReportMetrics.java]: Right

File: docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java (hunk 9 above)
Comment (lordofthejars) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Yes let's do it after Cube is merge, if not we are going to spend all the time trying to fix things of other extensions without releasing Cube, and for now we can live without that changes.

Comment (bartoszmajsak) [inline docker/recorder/src/test/java/org/arquillian/cube/docker/impl/client/reporter/TakeDockerEnvironmentTest.java]: Sure, that's what I was thinking about. Not as a part of this PR, but in general.

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java (hunk 4 above)
Comment (lordofthejars) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java]: Yes as I said I know it is not the best way in terms of code, but at least it is the best in terms of accuracy. I can try to put in higher priority the event which will not be the same times but really close.

Comment (lordofthejars) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/model/DockerCube.java]: But as you said when we start adding more and more metrics we could find another kind of integrations. This is a simple one. Maybe an integration with http://metrics.dropwizard.io/3.1.0/

**Comment by lordofthejars** [comment]
You can configure a custom directory too. We can put them in target/videos if you think it is better. In that case it was only a decision, but we can change
